### ISSUE TITLE
Tiling-based Inference

### ISSUE THREAD

**Issue by denghilbert** [naver/croco#17]
Hi,

I wonder why the inference results looks so different with images I took and images in dataset.
Bike:
![im0](https://github.com/naver/croco/assets/56004925/90529ec2-136b-423d-b84d-4509954063fd)
![im1](https://github.com/naver/croco/assets/56004925/bb2fa7d8-c43e-4ccf-819b-ea5e99f3fdb5)
![bike](https://github.com/naver/croco/assets/56004925/a487b2cb-e248-4ac8-bccf-3740c18627c4)
My images:
![grill](https://github.com/naver/croco/assets/56004925/64f18dd6-6cfc-45e3-9209-f9d7c26a5415)
![grill0](https://github.com/naver/croco/assets/56004925/c133aacd-ccb6-4f79-b808-2d501f2ddebe)
![grill1](https://github.com/naver/croco/assets/56004925/336d8b16-02e5-424a-ad33-13bc352d6805)

Bike looks so smooth but I just use the same setting for inference.

Best

**Comment by PhilippeWeinzaepfel** [comment]
Hi and thanks for your interest in our work,

This is most likely due to our tiling-based inference scheme. Since we perform inference on tiles, typically of 704px width, while assuming crops at the same location in both images, one limitation of our model is for large disparity values, which typically occur in case of a wide baseline and extremely high resolution images as it seems to be the case in your example. The matching area might not be visible bcause of that.

In such a scenario, one workaround is to downscale images. Alternatively, one can perform a first inference (possibly at lower resolution) to sample tiles more smartly in a second phase. This is something we have tried and mentioned in appendix B.5 of the paper: for each crop, we first predict disparities considering the same crop coordinates in the second image. In case of the presence of high disparity values, we perform a second inference using a shifted crop coordinates in the second image to ensure the matching area is present and update these high disparity values.

I hope that helps,
Best
Philippe

**Comment by PhilippeWeinzaepfel** [comment]
(1) In the bicycle example, despite the high resolution, the maximum predicted dispartiy is ~145 (and probably the maximum ground-truth disparity is in the same range). So predictions based on tiles are fine. In your other example, the disparity values look really high, in which case the naive tiling strategy of sampling the 352x704 crop in the second image at the same location as in the first image totally fails, leading to really poor prediction. When merging these really poor prediction made for each tile, it just output something totally messed up. It is about overfitting, it is a known failure mode of the tiling approach. For instance, this is the result on the Middlebury test set on the Hoops example where the disparity range is high on the pillar of the stairs, and it is also failing dramatically with tiling artifact highly visible in these areas.
![image](https://github.com/naver/croco/assets/13852085/6145e84e-2bc7-41fa-b01e-f3395d9499fb)

(2) Yes. During training, we take random crop of 352x704 in the first image, and consider the crop at the same location and size in the second image. At test time, we perform a sliding window with the same 352x704 window size in the first image and always consider the crop at the same localtion and size in the second image.

**Comment by denghilbert** [comment]
Thank you so much for this patient explanation.
